File: Stocks/StockModel.m
//
//  StockModel.m
//  Stocks
//
//  Created by Jonathan Wight on 6/10/14.
//  Copyright (c) 2014 schwa.io. All rights reserved.
//

#import "StockModel.h"

@interface StockModel ()
@property (readwrite, nonatomic, copy) NSArray *symbols;
@end

@implementation StockModel

static id gSharedInstance = NULL;

+ (instancetype)sharedInstance
{
    static dispatch_once_t sOnceToken = 0;
    dispatch_once(&sOnceToken, ^{
        gSharedInstance = [[self alloc] init];
    });
    return(gSharedInstance);
}


- (id)init
{
    if ((self = [super init]) != NULL)
    {
        _symbols = @[ @"TWTR", @"AAPL" ];
    }
    return self;
}

- (void)fetchInfoForSymbol:(NSString *)inSymbol handler:(void (^)(NSDictionary *info, NSError *error))inHandler
{
    NSString *theURLString = [NSString stringWithFormat:@"http://dev.markitondemand.com/Api/v2/Quote/json?symbol=%@", inSymbol];
    NSURL *theURL = [NSURL URLWithString:theURLString];
    NSURLRequest *theRequest = [NSURLRequest requestWithURL:theURL];

    NSURLSessionDataTask *theTask = [[NSURLSession sharedSession] dataTaskWithRequest:theRequest completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        // Properly handling HTTP errors and bad JSON is an excercise for someone less lazy.
        NSDictionary *theJSON = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
        inHandler(theJSON, NULL);
    }];
    [theTask resume];
}

@end

File: Stocks/StockModel.swift
//
//  StockModel.swift
//  Stocks
//
//  Created by Jim Kubicek on 6/10/14.
//  Copyright (c) 2014 schwa.io. All rights reserved.
//

import UIKit

let sharedStockModel = StockModel()

struct StockModel {

    let symbols = ["TWTR", "AAPL"]

    func fetchInfoForSymbol(symbol: String, handler: (info: Dictionary<String,Int>?, error: NSError?) -> ()) {

        let urlString = "http://dev.markitondemand.com/Api/v2/Quote/json?symbol=\(symbol)"
        let url = NSURL.URLWithString(urlString)


        let session = NSURLSession.sharedSession()
        let task = session.dataTaskWithURL(url) {
            (data, response, error) in
            var jsonError: NSError?
            let opt = NSJSONReadingOptions(0)
            let json = NSJSONSerialization.JSONObjectWithData(data, options: opt, error: &jsonError) as? Dictionary<String,Int>
            handler(info: json, error: jsonError)
        }
        task.resume()
    }
}
